plugins: add InitializationSample to test init samples

elgiano opened the pull request:
## Purpose and Motivation

Part of #6813, adds tests for all related PRs (#6810, #6816, #6832, #6822, #6805).
In order to test init samples, I wrote a new UGen `InitializationSample` that just output its input init sample forever (caching it at Ctor time).

## Types of changes

- New feature

## To-do list

- [x] Code is tested
- [ ] All tests are passing (see merge notes)
- [ ] Updated documentation
- [ ] This PR is ready for review

## Merge notes
- Dependencies: tests for Phasor and Sweep are failing before #6805 and #6822 are merged
- Missing documentation for InitializationSample

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+			if(ugenClass.respondsTo(\ar)) {
+				tests = tests ++ [
+					"{%.ar(DC.ar(0))}",
+					"{%.ar(DC.ar(1))}",
```
It would be good to use less common "default" values for the input signal. E.g. a problematic UGen might have it's init sample hard-coded to `0.0`, so the test signal `DC.ar(0)` wouldn't catch the bug.

elgiano replied:
the intention was to test triggers, but I see your point and we could use -1234 and 1234 instead

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+			if(ugenClass.respondsTo(\kr)) {
+				tests = tests ++ [
+					"{%.kr(DC.kr(0))}",
+					"{%.kr(DC.kr(1))}",
```
same as above wrt default input values

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	test_TriggerUGens {
+		var basicTest = { |ugenClass|
```
are these tests "basic" in the sense that they're only testing the init sample and with matched input/output rates?

maybe a more descriptive name to that effect, like `initSampleTest`

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+		this.assertInitEqualsFirstSample({Stepper.ar(DC.ar(1), DC.ar(1))});
+		this.assertInitEqualsFirstSample({Stepper.ar(DC.ar(1), DC.ar(0))});
+		this.assertInitEqualsFirstSample({Stepper.ar(DC.ar(0), DC.ar(1))});
+		this.assertInitEqualsFirstSample({Stepper.ar(DC.ar(0), DC.ar(0))});
```
minor formatting: spaces within { }

There are more occurrences below, maybe just do a pass over the full file for formatting

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+		this.assertInitEqualsFirstSample { InitializationSample.ar(0) }
+	}
+
+	test_TriggerUGens {
```
This is discretionary, but... this test method is feeling rather monolithic. 

I see why it can be useful to have tests grouped by UGen, but you group instead by the kind of test. One simple approach would be to have init==first sample tests in their own test method, and break out the `assertSynthOutput` tests to their own methods.

There may be clues how to more clearly separate tests by [naming convention](https://github.com/supercollider/supercollider/wiki/Unit-Testing-Guide#naming).

elgiano replied:
Yeah I always wonder about this, since splitting server tests involves extra overhead for server start and stop. But it seems reasonable to split initSample == y[0] from the others

elgiano replied:
I've split tests that use `assertInitEqualsFirstSample` from the ones using `assertSynthOutput`, which went to a separate `test_TriggerUGens_initialOutput` method

mtmccrea commented on testsuite/classlibrary/TestUGen_initSample.sc:
```diff
@@ -0,0 +1,253 @@
+		server.remove;
+	}
+
+	// Async helpers
```
Are these async helpers useful enough to promote to a superclass that all UGen unit tests might have access to?

mtmccrea reviewed (commented):
Thanks! This is nice factoring of many tests!

elgiano commented:
@mtmccrea I've finally got around to addressing all your comments :) thank you!
- I've refactored the big monolithic test in two separate ones: I wouldn't split it more than this because of the overhead of starting/stopping a separate server for each test method.
- Replaced input values 0 and 1 with less common -1234 and 1234
- fixed formatting

All the tests are passing.

Now, for me there are only two things left:
- Reusing async helpers in other test methods, but I think this should be another PR.
- In this block:
```supercollider
		// test start on reset when triggered at init
		this.assertSynthOutput({ Phasor.ar(1, 1, 0, 10, 5) }, [5]);
		// AK and AA subsample interpolation causes unexpected out: starts at 6
		this.assertSynthOutput({ Phasor.kr(1, 1, 0, 10, 5) }, [6]);
		this.assertSynthOutput({ Phasor.ar(DC.ar(1), 1, 0, 10, 5) }, [6]);
		this.assertSynthOutput({ Phasor.ar(DC.ar(1), DC.ar(1), 0, 10, 5) }, [6]);
```
Phasor resets to `reset + 1`, i.e. 6 in this case, for all except the first rates combination. We know this is a controversial behavior... I'm not sure whether it's better to test for it as it is, or not to test it at all.

mtmccrea commented:
Thanks for your work on this!
>Reusing async helpers in other test methods, but I think this should be another PR.

Agreed.

>Phasor resets to reset + 1, i.e. 6 in this case, for all except the first rates combination. We know this is a controversial behavior... I'm not sure whether it's better to test for it as it is, or not to test it at all.

I think it's ok to leave the tests, because once the Phasor behavior is fixed or otherwise resolved, these tests should break and it never hurts to consider why they break and update them :)

That said, could the explanatory comment be more specific, something as simple as:
>AK and AA subsample interpolation of the trigger causes unexpected out:

and also reference the appropriate Issue/PR?

elgiano commented:
Thanks! So I've left those tests and added a link to #6883 
Since I was there I noticed I had left a note about missing RunningMin/RunningMax tests, and I added them!

I think this is good to go!
